<a href="https://colab.research.google.com/github/ihertting/CS167_Notes/blob/main/Notebook2_starter.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Notebook \#2

**Name**: 



## Load in the data:
You'll notice that there are a lot of columns here. We're going to work with a subset of these columns, not the entire dataset.

In [6]:
import pandas
import numpy
from google.colab import drive 
drive.mount('/content/drive')
data = pandas.read_csv('/content/drive/MyDrive/datasets/vehicles.csv')
pandas.set_option('display.max_columns', 100)
data.head()

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


/usr/local/lib/python3.7/dist-packages/IPython/core/interactiveshell.py:2718: DtypeWarning: Columns (70,71,72,73,74,76,79) have mixed types.Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


barrels08  barrelsA08  charge120  charge240  city08  city08U  cityA08  \
0  15.695714         0.0        0.0        0.0      19      0.0        0   
1  29.964545         0.0        0.0        0.0       9      0.0        0   
2  12.207778         0.0        0.0        0.0      23      0.0        0   
3  29.964545         0.0        0.0        0.0      10      0.0        0   
4  17.347895         0.0        0.0        0.0      17      0.0        0   

   cityA08U  cityCD  cityE  cityUF  co2  co2A  co2TailpipeAGpm  \
0       0.0     0.0    0.0     0.0   -1    -1              0.0   
1       0.0     0.0    0.0     0.0   -1    -1              0.0   
2       0.0     0.0    0.0     0.0   -1    -1              0.0   
3       0.0     0.0    0.0     0.0   -1    -1              0.0   
4       0.0     0.0    0.0     0.0   -1    -1              0.0   

   co2TailpipeGpm  comb08  comb08U  combA08  combA08U  combE  combinedCD  \
0      423.190476      21      0.0        0       0.0    0.0         0.0   
1      807.909091      11      0.0        0       0.0    0.0         0.0   
2      329.148148      27      0.0        0       0.0    0.0         0.0   
3      807.909091      11      0.0        0       0.0    0.0         0.0   
4      467.736842      19      0.0        0       0.0    0.0         0.0   

   combinedUF  cylinders  displ                       drive  engId  \
0         0.0        4.0    2.0            Rear-Wheel Drive   9011   
1         0.0       12.0    4.9            Rear-Wheel Drive  22020   
2         0.0        4.0    2.2           Front-Wheel Drive   2100   
3         0.0        8.0    5.2            Rear-Wheel Drive   2850   
4         0.0        4.0    2.2  4-Wheel or All-Wheel Drive  66031   

     eng_dscr  feScore  fuelCost08  fuelCostA08 fuelType         fuelType1  \
0       (FFS)       -1        1850            0  Regular  Regular Gasoline   
1   (GUZZLER)       -1        3500            0  Regular  Regular Gasoline   
2       (FFS)       -1        1450            0  Regular  Regular Gasoline   
3         NaN       -1        3500            0  Regular  Regular Gasoline   
4  (FFS,TRBO)       -1        2550            0  Premium  Premium Gasoline   

   ghgScore  ghgScoreA  highway08  highway08U  highwayA08  highwayA08U  \
0        -1         -1         25         0.0           0          0.0   
1        -1         -1         14         0.0           0          0.0   
2        -1         -1         33         0.0           0          0.0   
3        -1         -1         12         0.0           0          0.0   
4        -1         -1         23         0.0           0          0.0   

   highwayCD  highwayE  highwayUF  hlv  hpv     id  lv2  lv4        make  \
0        0.0       0.0        0.0    0    0      1    0    0  Alfa Romeo   
1        0.0       0.0        0.0    0    0     10    0    0     Ferrari   
2        0.0       0.0        0.0   19   77    100    0    0       Dodge   
3        0.0       0.0        0.0    0    0   1000    0    0       Dodge   
4        0.0       0.0        0.0    0    0  10000    0   14      Subaru   

                 model mpgData  phevBlended  pv2  pv4  range  rangeCity  \
0   Spider Veloce 2000       Y        False    0    0      0        0.0   
1           Testarossa       N        False    0    0      0        0.0   
2              Charger       Y        False    0    0      0        0.0   
3  B150/B250 Wagon 2WD       N        False    0    0      0        0.0   
4     Legacy AWD Turbo       N        False    0   90      0        0.0   

   rangeCityA  rangeHwy  rangeHwyA            trany    UCity  UCityA  \
0         0.0       0.0        0.0     Manual 5-spd  23.3333     0.0   
1         0.0       0.0        0.0     Manual 5-spd  11.0000     0.0   
2         0.0       0.0        0.0     Manual 5-spd  29.0000     0.0   
3         0.0       0.0        0.0  Automatic 3-spd  12.2222     0.0   
4         0.0       0.0        0.0     Manual 5-spd  21.0000     0.0   

   UHighway  UHi

## Exercise #1: 
**Work with the right subset**: You're not going to work with the whole data set, just the cars with `"Regular"` listed in the `"fuelType"` column. So, to start, make a subset of the original data.

In [113]:
# your code here:
subset = data[data['fuelType'] == 'Regular'][['year', 'cylinders', 'displ', 'comb08']]
subset

year  cylinders  displ  comb08
0      1985        4.0    2.0      21
1      1985       12.0    4.9      11
2      1985        4.0    2.2      27
3      1985        8.0    5.2      11
5      1993        4.0    1.8      22
...     ...        ...    ...     ...
42097  1993        4.0    1.9      25
42098  1993        4.0    2.2      22
42099  1993        4.0    2.2      23
42100  1993        4.0    2.2      21
42101  1993        4.0    2.2      21

[26863 rows x 4 columns]

## Exercise #2: 
**Check for null values**: If you have a null target value, you will need to throw that example out. 

If you have a null predictor value, you can either fill them in with something (like the mean/median) or you can drop those rows from the data set. Useful functions here are `isna()`, `any()`, `fillna()`, `value_counts()` and `dropna()`. If you're not sure what these do, look them up in the [pandas documentation](https://pandas.pydata.org/docs/).

If you've tried the functions above, but your data still has missing values, you probably need to do one of the following thigns:
-  Assign the results of the above functions to a variable, `clean_data = original_data.dropna()`,
- use the `inplace=True` parameter when you call the above functions.

Describe in a markup cell what you decided to do with the null data and why.

In [124]:
## remove missing data here

subset.isna().any()

cylinders_mean = subset['cylinders'].mean()
displ_mean = subset['displ'].mean()

subset['cylinders'].fillna(cylinders_mean, inplace=True)
subset['displ'].fillna(displ_mean, inplace=True)

subset.isna().any()

year         False
cylinders    False
displ        False
comb08       False
dtype: bool

### Describe what you did:
Because the only null values were predictor values, I did not have to throw any examples out. So, I replaced the missing values with the the average value for the respective column.


## Exercise #3:
Write up a k-nearest-neighbors function like the one you made for the iris data set in class. It should be able to make mpg ("comb08") predictions for new cars based on the year, cylinders, and displacement. You should also be able to specify what you want to use as k.

Pay special attention to whether this problem is a classification or a regression, and how the answer to that question may change the output of your model. 

In [128]:
## Build your function here
def knn(car, data, k):

  data['distance'] = numpy.sqrt(
    (car['year']-data['year'])**2
    +(car['cylinders']-data['cylinders'])**2
    +(car['displ']-data['displ'])**2)

  sorted_data = data.sort_values(['distance'])

  return sorted_data.head(k)['comb08'].mean()

## Exercise #4: 
Demonstrate that your function works by making up some new values for hypothetical cars and using your function to display the predicted mpg ("comb08") for that car. 

In [131]:
## demonstrate that your function works here

car1 = {}
car1['year'] = 1985
car1['cylinders'] = 4
car1['displ'] = 5

car2 = {}
car2['year'] = 1993
car2['cylinders'] = 8
car2['displ'] = 2

print("First car mpg: ", knn(car1, subset, 5))
print("Second car mpg: ",knn(car2, subset, 3))

First car mpg:  15.6
Second car mpg:  20.0


## Exercise #5
Make a copy of the data and normalize the training data using Z-score-- `copy()` will be useful here. Predict the mpg of a particular car using your k-nearest-neighbors function with both the normalized and non-normalized training data. Compare your results. Use a markup cell to describe and explain the differences in a few sentences.

In [134]:
## normalize a copy of the data

# copy the data
norm_subset = subset.copy()

# means
year_mean = subset['year'].mean()
cylinders_mean = subset['cylinders'].mean()
displ_mean = subset['displ'].mean()

# stds
year_std = subset['year'].std()
cylinders_std = subset['cylinders'].std()
displ_std = subset['displ'].std()

# replacing the data with z scores
norm_subset['year'] = (norm_subset['year'] - year_mean)/year_std
norm_subset['cylinders'] = (norm_subset['cylinders'] - cylinders_mean)/cylinders_std
norm_subset['displ'] = (norm_subset['displ'] - displ_mean)/displ_std

# non-normalized and normalized car
car = {}
car['year'] = 1985
car['cylinders'] = 4
car['displ'] = 5

norm_car = {}
norm_car['year'] = (1985 - year_mean)/year_std
norm_car['cylinders'] = (4- cylinders_mean)/cylinders_std
norm_car['displ'] = (5 - displ_mean)/displ_std

print('non-normalized knn: ', knn(car, subset, 5))
print('normalized knn: ', knn(norm_car, norm_subset, 5))

non-normalized knn:  15.6
normalized knn:  15.0


## Describe and explain the differences:
The non-normalized fifth nearest neighbor prediction is 15.6 mpg, whereas the normalized fifth nearest neighbor prediction is a bit lower at 15 mpg.

The normalized data was with a Z-score. A Z-score is a standardized value that compares the individual point with the mean of the set. This way, the normalized prediction is likely a bit more accurate because you could look at the dataset it came from as a level playing field, in a sense.